find_files only from loaded_gems when using gemdeps

mpapis opened the pull request:
when working in context of `RUBYGEMS_GEMDEPS=-` - rubygems will not allow loading all gems - only those in found `Gemfile` are allowed, this PR restricts the list in `find_files` - to the allowed gems

segiddins commented:
This check seems to be necessary in a few places -- why not just change `Gem::Specification.stubs` ?

mpapis commented:
@segiddins "fixing" `stubs` would break kernel patch for `require` in a case when:

1. load ruby in a path without `Gemfile`
2. load a gem from installed gems and `require` it
3. run `Gem.use_gemdeps "path/to/Gemfile"` (without the gem in line 2)
4. `require` the same gem from point 2 again

krainboltgreene commented:
This looks good to me :+1:.

segiddins commented:
It's missing test coverage?

krainboltgreene commented:
Yep, looks to be missing that.

mpapis commented:
rebased and added test => https://github.com/rubygems/rubygems/pull/1277/files#diff-138611698f18248c77ff83e3b8165507R499 - this is the difference from the above tests - works as expected

mpapis commented:
and the tests passed finally, I had to disable `1.8.7` because of `$SAFE=1` in tests was triggering `SecurityError: Insecure operation - mkdir` via `write_file` used to write the `Gemfile`, this problem is only test related and should not influence the fix itself

krainboltgreene commented:
:+1:

duckinator commented:
This looks good to me, too. :+1:
